pkg/azure/provider/helpers/driver.go

```diff
@@ -0,0 +1,675 @@
+func SkipDeleteMachine(ctx context.Context, factory access.Factory, connectConfig access.ConnectConfig, resourceGroup string) (bool, error) {
+	resGroupAccess, err := factory.GetResourceGroupsAccess(connectConfig)
+	if err != nil {
+		return false, status.Error(codes.Internal, fmt.Sprintf("failed to create ResourceGroup access to process request: [ResourceGroup: %s]", resourceGroup))
+	}
+	resGroupExists, err := accesshelpers.ResourceGroupExists(ctx, resGroupAccess, resourceGroup)
+	if err != nil {
+		return false, status.Error(codes.Internal, fmt.Sprintf("failed to check if ResourceGroup %s exists, Err: %v", resourceGroup, err))
+	}
+	return !resGroupExists, nil
+}
+
+// GetDiskNames creates disk names for all configured OSDisk and DataDisk in the provider spec.
+func GetDiskNames(providerSpec api.AzureProviderSpec, vmName string) []string {
+	dataDisks := providerSpec.Properties.StorageProfile.DataDisks
+	diskNames := make([]string, 0, len(dataDisks)+1)
+	diskNames = append(diskNames, utils.CreateOSDiskName(vmName))
+	if !utils.IsSliceNilOrEmpty(dataDisks) {
+		for _, disk := range dataDisks {
+			diskName := utils.CreateDataDiskName(vmName, disk)
+			diskNames = append(diskNames, diskName)
+		}
+	}
+	return diskNames
+}
+
+// CheckAndDeleteLeftoverNICsAndDisks creates tasks for NIC and DISK deletion and runs them concurrently. It waits for them to complete and then returns a consolidated error if there is any.
+// This method will be called when these resources are left without an associated VM.
+func CheckAndDeleteLeftoverNICsAndDisks(ctx context.Context, factory access.Factory, vmName string, connectConfig access.ConnectConfig, providerSpec api.AzureProviderSpec) error {
+	// Gather the names for NIC, OSDisk and Data Disks that needs to be checked for existence and then deleted if they exist.
+	resourceGroup := providerSpec.ResourceGroup
+	nicName := utils.CreateNICName(vmName)
+	diskNames := GetDiskNames(providerSpec, vmName)
+
+	// create NIC and Disks clients
+	nicAccess, err := factory.GetNetworkInterfacesAccess(connectConfig)
+	if err != nil {
+		return status.Error(codes.Internal, fmt.Sprintf("Failed to create nic access for VM: [ResourceGroup: %s, Name: %s], Err: %v", resourceGroup, vmName, err))
+	}
+	disksAccess, err := factory.GetDisksAccess(connectConfig)
+	if err != nil {
+		return status.Error(codes.Internal, fmt.Sprintf("Failed to create disk access for VM: [ResourceGroup: %s, Name: %s], Err: %v", resourceGroup, vmName, err))
+	}
+
+	// Create NIC and Disk deletion tasks and run them concurrently.
+	tasks := make([]utils.Task, 0, len(diskNames)+1)
+	tasks = append(tasks, createNICDeleteTask(resourceGroup, nicName, nicAccess))
+	//tasks = append(tasks, d.createDiskDeletionTasks(resourceGroup, diskNames, disksAccess)...)
+	tasks = append(tasks, createDisksDeletionTask(resourceGroup, diskNames, disksAccess))
```

Comment:
Currently we have taken an approach which is conservative. We only delete resources (grouped by type) concurrently. This means if there are 1 NIC and 10 Disks then - NIC and Disk deletion will happen concurrently but the task that deletes Disks will delete one Disk at a time.

This is a bit conservative knowing the following:
1. There is no limit to the number of calls to `DeleteMachine` that can be made from MCM
2.  Azure has historically had problems with disk attach and detach. I am not sure if Deletion will cause an issue.
3. During rolling update many organizations within SAP have overlapping maintenance schedules thus causing a LOT of requests to Azure. 

This approach will add an additional latency to the `DeleteMachine` operation but will be less prone to issues from Azure that are caused by too many requests for Disk deletion. Since we do not have any limit on MCM side we put some bounds on how difference resources are deleted across VMs in the provider code.

Reply:
We propose to use the library differently:
* have a separate task for each resource (nic/disk) , and have a bound of 2 (=1 if we want to keep it serial), as with the current way, its parallel in first step, and then serial , which essentially can be achieved without the library.

The above proposed way will allow to make `bound` configurable in the future, without much changes in the code.

Reply:
Changed. PTAL